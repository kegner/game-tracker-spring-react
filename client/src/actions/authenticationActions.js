export const SET_USER = "SET_USER";
export const CLEAR_USER = "CLEAR_USER";

export const setUserData = (user) => ({
  type: SET_USER,
  payload: { user }
});

export const clearUserData = () => ({
    type: CLEAR_USER
});