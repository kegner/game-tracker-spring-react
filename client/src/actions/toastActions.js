export const SHOW_SUCCESS = "SHOW_SUCCESS";
export const SHOW_ERROR = "SHOW_ERROR";
export const HIDE_TOAST = "HIDE_TOAST";

export const showSuccessToast = (message) => ({
  type: SHOW_SUCCESS,
  payload: { message }
});

export const showErrorToast = (message) => ({
  type: SHOW_ERROR,
  payload: { message }
});

export const hideToast = (id) => ({
  type: HIDE_TOAST,
  payload: { id }
});