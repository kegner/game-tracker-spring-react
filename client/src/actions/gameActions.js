import axios from 'axios';
import { showSuccessToast, showErrorToast } from './toastActions';

export const SELECT_GAME = "SELECT_GAME";
export const FETCH_GAMES = "FETCH_GAMES";
export const CREATE_GAME = "CREATE_GAME";
export const UPDATE_GAME = "UPDATE_GAME";
export const DELETE_GAME = "DELETE_GAME";
export const START_LOADING = "START_LOADING";
export const START_SAVING = "START_SAVING";
export const STOP_LOADING = "STOP_LOADING";
export const STOP_SAVING = "STOP_SAVING";

const fetchSuccess = (games) => ({
  type: FETCH_GAMES,
  payload: { games }
});

const createSuccess = (game) => ({
  type: CREATE_GAME,
  payload: { game }
});

const updateSuccess = (game) => ({
  type: UPDATE_GAME,
  payload: { game }
});

const deleteSuccess = (id) => ({
  type: DELETE_GAME,
  payload: { id }
});

const startLoading = () => ({
  type: START_LOADING
});

const startSaving = () => ({
  type: START_SAVING
});

const stopLoading = () => ({
  type: STOP_LOADING
});

const stopSaving = () => ({
  type: STOP_SAVING
});

const errorHandler = (error, dispatch) => {
  const validResponses = [404, 409];
  if (error && error.response && validResponses.includes(error.response.status)) {
    // this is the expected error format for Spring exception messages
    dispatch(showErrorToast(error.response.data.message));
  } else {
    console.error(error);
  }
};

export const selectGame = (game) => ({
  type: SELECT_GAME,
  payload: { game }
});

export const fetchGames = (userId) => async (dispatch) => {
  dispatch(startLoading());

  try {
    const res = await axios.get("/api/v1/games", { params: { userId } });
    dispatch(fetchSuccess(res.data.games));
  } catch (error) {
    dispatch(stopLoading());
    errorHandler(error, dispatch);
  }
};

export const createGame = (game) => async (dispatch) => {
  dispatch(startSaving());

  try {
    const res = await axios.post("/api/v1/games", game);

    dispatch(createSuccess(res.data));
    dispatch(showSuccessToast("Game added."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};

export const updateGame = (game) => async (dispatch) => {
  dispatch(startSaving());

  try {
    const res = await axios.put("/api/v1/games", game);

    dispatch(updateSuccess(res.data));
    dispatch(showSuccessToast("Game updated."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};

export const deleteGame = (id) => async (dispatch) => {
  dispatch(startSaving());

  try {
    await axios.delete(`/api/v1/games/${id}`);
    dispatch(deleteSuccess(id));
    dispatch(showSuccessToast("Game deleted."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};