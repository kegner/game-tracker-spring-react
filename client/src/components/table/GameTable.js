import React, { useEffect, useState, useRef } from 'react'
import { useSelector, useDispatch } from 'react-redux';
import { fetchGames } from '../../actions/gameActions';
import Spinner from '../layout/Spinner';
import paginate from '../../utils/paginate';
import TableBody from './TableBody';
import TableFooter from './TableFooter';

const GameTable = () => {

  const itemsPerPage = 5;

  const [pageIndex, setPageIndex] = useState(0);
  const [gamesPages, setGamesPages] = useState([[]]);

  const { games, loading } = useSelector(state => state.gameReducer);
  const { user } = useSelector(state => state.authenticationReducer)
  const previousCount = useRef(0);
  
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchGames(user.id));
  }, [user.id, dispatch]);

  useEffect(() => {
    if (previousCount.current === 0) {
      previousCount.current = games.length;
    } else {
      if (previousCount.current < games.length) { // add a game
        setPageIndex(Math.floor(Math.max(games.length - 1, 0) / itemsPerPage));
      } else if (previousCount.current > games.length) { // delete a game
        setPageIndex(index => Math.min(Math.floor(Math.max(games.length - 1, 0) / itemsPerPage), index));
      }
      previousCount.current = games.length;
    }
    setGamesPages(paginate(games, itemsPerPage));
  }, [games]);

  const spinner = loading ? <Spinner /> : <></>;

  return (
    <div className="card">
      <div className="card-header bg-dark text-white">
        Game Table
      </div>
      <div className="card-body p-0">
        <table className="table table-hover table-sm">
          <caption className="visually-hidden">Game Table</caption>
          <thead>
            <tr>
              <th scope="col" className="text-center">Title</th>
              <th scope="col" className="text-center">Status</th>
              <th scope="col" className="text-center">Rating</th>
              <th scope="col" className="text-center">Delete</th>
            </tr>
          </thead>
          <TableBody gamesPages={gamesPages} pageIndex={pageIndex} />
          <TableFooter gamesPages={gamesPages} pageIndex={pageIndex}
              setPageIndex={setPageIndex} itemsPerPage={itemsPerPage} />
        </table>
      </div>
      {spinner}
    </div>
  )
}

export default GameTable
