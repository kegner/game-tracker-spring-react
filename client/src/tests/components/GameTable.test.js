import React from 'react';
import { Provider } from "react-redux";
import { render } from '@testing-library/react';
import { FETCH_GAMES, START_LOADING } from '../../actions/gameActions';
import { createTestStore } from '../store/testStore';
import axios from 'axios';
import GameTable from '../../components/table/GameTable';

let store;
let dispatch;
let component;
let container;
let asFragment;
let games = [
  { id: "1", title: "Game 1" }
];

jest.mock("axios");

describe("GameTable", () => {

  beforeEach(() => {
    axios.get.mockResolvedValue({
      data: { games }
    });
    store = createTestStore({
      mockStore: true,
      initialState: {
        gameReducer: {
          games,
          selectedGame: games[0]
        },
        authenticationReducer: {
          user: {
            id: 1
          }
        }
      }
    });
    dispatch = store.dispatch;
    component = render(
      <Provider store={store}>
        <GameTable />
      </Provider>
    );
    container = component.container;
    asFragment = component.asFragment;
  });

  it("exists", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("fetches games", async () => {
    const expectedActions = [
      { type: START_LOADING },
      { type: FETCH_GAMES, payload: { games } }
    ];

    // await for all dispatches
    await new Promise(r => setTimeout(r, 1));
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(store.getActions()).toEqual(expectedActions);
  });

});