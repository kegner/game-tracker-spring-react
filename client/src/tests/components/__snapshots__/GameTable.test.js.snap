// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameTable exists 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card-header bg-dark text-white"
    >
      Game Table
    </div>
    <div
      class="card-body p-0"
    >
      <table
        class="table table-hover table-sm"
      >
        <caption
          class="visually-hidden"
        >
          Game Table
        </caption>
        <thead>
          <tr>
            <th
              class="text-center"
              scope="col"
            >
              Title
            </th>
            <th
              class="text-center"
              scope="col"
            >
              Status
            </th>
            <th
              class="text-center"
              scope="col"
            >
              Rating
            </th>
            <th
              class="text-center"
              scope="col"
            >
              Delete
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="table-active"
          >
            <td
              class="text-center"
            >
              Game 1
            </td>
            <td
              class="text-center"
            />
            <td
              class="text-center"
            />
            <td
              class="text-center"
            >
              <button
                aria-label="Delete"
                class="btn btn-danger"
              >
                <i
                  class="bi bi-trash"
                />
              </button>
            </td>
          </tr>
        </tbody>
        <tfoot>
          <tr>
            <td
              class="border-0"
              colspan="4"
            >
              <div
                class="d-flex"
              >
                <div
                  class="align-self-center ms-1"
                >
                  Showing 1 to 1 of 1
                </div>
                <div
                  class="ms-auto"
                >
                  <nav
                    aria-label="Page navigation example"
                  >
                    <ul
                      class="pagination my-1"
                    >
                      <li
                        class="page-item"
                      >
                        <button
                          class="page-link"
                          disabled=""
                        >
                          Previous
                        </button>
                      </li>
                      <li
                        class="page-item  active"
                      >
                        <button
                          class="page-link"
                        >
                          1
                        </button>
                      </li>
                      <li
                        class="page-item"
                      >
                        <button
                          class="page-link"
                          disabled=""
                        >
                          Next
                        </button>
                      </li>
                    </ul>
                  </nav>
                </div>
              </div>
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
